core/module.h

```diff
@@ -465,20 +505,113 @@ inline void Module::RunChooseModule(gate_idx_t ogate_idx,
     deadend(batch);
     return;
   }
+
   for (auto &hook : ogate->hooks()) {
     hook->ProcessBatch(batch);
   }
 
-  for (auto &hook : ogate->igate()->hooks()) {
-    hook->ProcessBatch(batch);
+  task->AddToRun(ogate->igate(), batch);
+}
+
+inline void Module::RunNextModule(const Task *task, bess::PacketBatch *batch) {
+  RunChooseModule(task, 0, batch);
+}
+
+inline void Module::DropPacket(const Task *task, bess::Packet *pkt) {
+  task->dead_batch()->add(pkt);
+  if (static_cast<size_t>(task->dead_batch()->cnt()) >=
+      bess::PacketBatch::kMaxBurst) {
+    deadend(task->dead_batch());
   }
+}
 
-  ctx.set_current_igate(ogate->igate_idx());
-  (static_cast<Module *>(ogate->next()))->ProcessBatch(batch);
+inline void Module::EmitPacket(const Task *task, bess::Packet *pkt,
+                               gate_idx_t ogate_idx) {
+  // Check if valid ogate is set
+  if (unlikely((ogates_.size() <= ogate_idx) || (!ogates_[ogate_idx]))) {
+    DropPacket(task, pkt);
+    return;
+  }
+
+  // Put a packet into the ogate
+  bess::OGate *ogate = ogates_[ogate_idx];
+  bess::PacketBatch *batch = task->get_gate_batch(ogate);
+  if (!batch) {
```

Comment:
You can rearrange the code like this
```c++
   if (!batch) {
     if (!ogate->hooks().empty()) {
       // Having separate batch to run ogate hooks
      batch = task->AllocPacketBatch();
      task->set_gate_batch(ogate, batch);
       gate_with_hook_[gate_with_hook_cnt_++] = ogate_idx;
     } else {
       // If no ogate hooks, just use next igate batch
      batch = task->get_gate_batch(igate);
      if (batch == nullptr) {
        batch = task->AllocPacketBatch();
        task->AddToRun(igate, batch);
        task->set_gate_batch(ogate, batch);
       }
       gate_without_hook_[gate_without_hook_cnt_++] = ogate_idx;
     }
   }
```
to eliminate use of temporary variable and make it easier to see when the `batch` pointer is assigned.